apis/v1alpha1/adminnetworkpolicy_types.go

```diff
@@ -0,0 +1,154 @@
+)
+
+//+kubebuilder:object:root=true
+//+kubebuilder:subresource:status
+
+// AdminNetworkPolicy is the Schema for the adminnetworkpolicies API
+type AdminNetworkPolicy struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata"`
+
+	// Specification of the desired behavior of AdminNetworkPolicy.
+	Spec AdminNetworkPolicySpec `json:"spec"`
+
+	// Status is the status to be reported by the implementation.
+	// +optional
+	Status AdminNetworkPolicyStatus `json:"status,omitempty"`
+}
+
+// AdminNetworkPolicySpec defines the desired state of AdminNetworkPolicy
+type AdminNetworkPolicySpec struct {
+	// Priority is an int32 value bound to 0 - 1000, the lowest priority,
+	// "0" corresponds to the highest importance, while higher priorities have
+	// lower importance.
+	// +kubebuilder:validation:Minimum=0
+	// +kubebuilder:validation:Maximum=1000
+	Priority int32 `json:"priority"`
+
+	// Subject defines the pods to which this AdminNetworkPolicy applies.
+	Subject AdminNetworkPolicySubject `json:"subject"`
+
+	// List of Ingress rules to be applied to the selected objects BEFORE any
+	// NetworkPolicy or BaslineAdminNetworkPolicy rules have been applied.
+	// A total of 100 rules will be allowed per each ANP instance.
+	// ANPs with no ingress rules do not affect ingress traffic.
+	// +optional
+	// +kubebuilder:validation:MaxItems=100
+	Ingress []AdminNetworkPolicyIngressRule `json:"ingress,omitempty"`
+
+	// List of Egress rules to be applied to the selected objects BEFORE any
+	// NetworkPolicy or BaslineAdminNetworkPolicy rules have been applied.
+	// A total of 100 rules will be allowed per each ANP instance.
+	// ANPs with no egress rules do not affect egress traffic.
+	// +optional
+	// +kubebuilder:validation:MaxItems=100
+	Egress []AdminNetworkPolicyEgressRule `json:"egress,omitempty"`
+}
+
+// AdminNetworkPolicyIngressRule describes an action to take on a particular
+// set of traffic destined for pods selected by an AdminNetworkPolicy's
+// Subject field. The traffic must match both ports and from.
+type AdminNetworkPolicyIngressRule struct {
+	// Name is an identifier for this rule, that should be no more than 100 characters
+	// in length.
+	// +optional
+	// +kubebuilder:validation:MaxLength=100
+	Name string `json:"name,omitempty"`
+
+	// Action specifies whether this rule must pass, allow or deny traffic.
+	// Allow: allows the selected traffic
+	// Deny: denies the selected traffic
+	// Pass: instructs the selected traffic to skip any remaining ANP rules, and
+	// then pass execution to any NetworkPolicies that select the pod.
+	// If the pod is not selected by any NetworkPolicies then execution
+	// is passed to any BaselineAdminNetworkPolicies that select the pod.
+	// This field is mandatory.
+	Action AdminNetworkPolicyRuleAction `json:"action"`
+
+	// Ports allows for matching on traffic based on port and protocols.
+	// This field is mandatory.
+	Ports AdminNetworkPolicyPorts `json:"ports"`
```

Comment:
After discussion, this wants to be a pointer, right?  If not specified, it covers all ports on all protocols.

Reply:
This is now just the list of AdminNetworkPolicyPort defined in each gress rule, i.e an empty of missing list means don't filter via port